http/apache/nanny.py

```diff
@@ -0,0 +1,337 @@
+#! /usr/bin/env python3
+"""Load balancer around apache."""
+import argparse
+import asyncio
+import concurrent.futures
+import dataclasses
+import functools
+import http.client
+import itertools
+import json
+import logging
+import os
+import pathlib
+import socket
+import shutil
+import subprocess
+from typing import Collection, Generic, List, TypeVar, AsyncContextManager
+
+import psutil
+
+T = TypeVar('T')
+
+logger = logging.getLogger()
+
+APACHE_RUN_DIR = '/var/run/apache2'
+APACHE_LOG_DIR = '/var/log/apache2'
+APACHE_LOCK_DIR = '/var/lock/apache2'
+
+
+@dataclasses.dataclass
+class ApacheInstance:
+    port: int
+    log_dir: pathlib.Path
+    pid_file_path: pathlib.Path
+
+
+class _ManagedResource(AsyncContextManager[T]):
+    _queue: "asyncio.Queue[T]"
+    _resource: T
+
+    def __init__(self, queue: "asyncio.Queue[T]"):
+        self._queue = queue
+
+    async def __aenter__(self) -> T:
+        self._resource = await self._queue.get()
+        return self._resource
+
+    async def __aexit__(self, *args, **kwargs):
+        self._queue.put_nowait(self._resource)
+
+
+class ResourcePool(Generic[T]):
+    """Asynchronous resource pool."""
+
+    _queue: "asyncio.Queue[T]"
+
+    def __init__(self, resources: Collection[T]):
+        self._queue = asyncio.Queue(len(resources))
+        for resource in resources:
+            self._queue.put_nowait(resource)
+
+    def borrow(self) -> _ManagedResource[T]:
+        """Return an async context manager allowing access to a resource."""
+        return _ManagedResource(self._queue)
+
+
+class NonClosingResponse(http.client.HTTPResponse):
+    def _close_conn(self):
+        # Normally this would close the file obj, but since we're using
+        # pipelining, we need to read multiple responses from the same
+        # connection.
+        pass
+
+
+class MockSocket:
+    """Mocket.
+
+    HTTPResponse only uses the ``makefile`` method of sockets.
+    """
+    def __init__(self, fp):
+        self.fp = fp
+    def makefile(self, *args, **kwargs):
+        return self.fp
+
+
+async def read_with_prepended_length(reader: asyncio.StreamReader) -> bytes:
+    """Raises IncompleteReadError if not all data is found."""
+    expected_size_str = await reader.readuntil(b'\n')
+    expected_size = int(expected_size_str)
+    return await reader.readexactly(expected_size)
+
+
+async def handle_request_stream(
```

Comment:
Right now, this function doesn't do a good enough job of finishing before `parser.py`'s timeout is up. This is due to having too much async stuff. Asyncio doesn't prioritize continuing older coroutines, thus newer requests sometimes get handled before finishing up old ones. We _do_ need all of these operations to use non-blocking I/O, though, since we expect apache to time out in some cases and can't be blocked on waiting for that.

I'd like to go back through this and find a solution. It's still appropriate to use async, just not in this way/to this extent (one low-hanging place where async can be excised: returning the apache instance to the pool is async, but could be made synchronous).

As for whether this should be a blocker for this PR, I'm not sure. Opinions appreciated. I'd at least like to make the parenthesized change above and see how effective it is.

Reply:
After removing some of the async stuff in this file, I haven't seen any really suspicious timeouts anymore. Worth keeping an eye on this if we start to see timeouts in the future, but for now it's in a good state.